extensions/agroal/runtime/src/main/java/io/quarkus/agroal/metrics/AgroalCounter.java

```diff
@@ -0,0 +1,98 @@
+package io.quarkus.agroal.metrics;
+
+import org.eclipse.microprofile.metrics.Counter;
+
+import io.agroal.api.AgroalDataSource;
+import io.agroal.api.AgroalDataSourceMetrics;
+import io.quarkus.arc.Arc;
+
+public class AgroalCounter implements Counter {
+
+    private String dataSourceName;
+    private volatile AgroalDataSource dataSource;
+    private String metric;
+
+    public AgroalCounter() {
+
+    }
+
+    /**
+     * @param dataSourceName Which datasource should be queried for metric
+     * @param metricName Name of the method from DataSource.getMetrics() that should be called to retrieve the particular value.
+     *        This has nothing to do with the metric name from MP Metrics point of view!
+     */
+    public AgroalCounter(String dataSourceName, String metricName) {
+        this.dataSourceName = dataSourceName;
+        this.metric = metricName;
+    }
+
+    public String getDataSourceName() {
+        return dataSourceName;
+    }
+
+    public void setDataSourceName(String dataSourceName) {
+        this.dataSourceName = dataSourceName;
+    }
+
+    private AgroalDataSource getDataSource() {
+        AgroalDataSource dsLocal = dataSource;
```

Comment:
It's ot mandatory to store the attribute inside a local variable. Volatile is enought to make it correct.
See the last example on https://www.cs.umd.edu/~pugh/java/memoryModel/DoubleCheckedLocking.html

Reply:
I got this from the interview with Joshua Bloch https://www.oracle.com/technical-resources/articles/javase/bloch-effective-08-qa.html, also the book Effective Java says that thanks to the local variable you get about a performance gain, because you only need to access the volatile field once if it's already non-null.  From wikipedia:
> Note the local variable "localRef", which seems unnecessary. The effect of this is that in cases where helper is already initialized (i.e., most of the time), the volatile field is only accessed once (due to "return localRef;" instead of "return helper;"), which can improve the method's overall performance by as much as 25 percent.

Reply:
I never thought about these 2 volatile reads, I really need to read again Effective Java ;)
Thanks for the pointers !